codex/blockexchange/engine/engine.nim

```diff
@@ -100,124 +93,147 @@ type
+  if self.blockexcRunning:
     warn "Starting blockexc twice"
     return
 
-  b.blockexcRunning = true
-  for i in 0 ..< b.concurrentTasks:
-    let fut = b.blockexcTaskRunner()
-    b.trackedFutures.track(fut)
-    asyncSpawn fut
+  self.blockexcRunning = true
+  for i in 0 ..< self.concurrentTasks:
+    let fut = self.blockexcTaskRunner()
+    self.trackedFutures.track(fut)
 
-proc stop*(b: BlockExcEngine) {.async.} =
+proc stop*(self: BlockExcEngine) {.async.} =
   ## Stop the blockexc blockexc
   ##
 
-  await b.discovery.stop()
-  await b.advertiser.stop()
+  await self.trackedFutures.cancelTracked()
+  await self.network.stop()
+  await self.discovery.stop()
+  await self.advertiser.stop()
 
   trace "NetworkStore stop"
-  if not b.blockexcRunning:
+  if not self.blockexcRunning:
     warn "Stopping blockexc without starting it"
     return
 
-  b.blockexcRunning = false
-  await b.trackedFutures.cancelTracked()
+  self.blockexcRunning = false
 
   trace "NetworkStore stopped"
 
 proc sendWantHave(
-    b: BlockExcEngine, addresses: seq[BlockAddress], peers: seq[BlockExcPeerCtx]
+    self: BlockExcEngine, addresses: seq[BlockAddress], peers: seq[BlockExcPeerCtx]
 ): Future[void] {.async.} =
   for p in peers:
     let toAsk = addresses.filterIt(it notin p.peerHave)
     trace "Sending wantHave request", toAsk, peer = p.id
-    await b.network.request.sendWantList(p.id, toAsk, wantType = WantType.WantHave)
+    await self.network.request.sendWantList(p.id, toAsk, wantType = WantType.WantHave)
     codex_block_exchange_want_have_lists_sent.inc()
 
 proc sendWantBlock(
-    b: BlockExcEngine, addresses: seq[BlockAddress], blockPeer: BlockExcPeerCtx
+    self: BlockExcEngine, addresses: seq[BlockAddress], blockPeer: BlockExcPeerCtx
 ): Future[void] {.async.} =
   trace "Sending wantBlock request to", addresses, peer = blockPeer.id
-  await b.network.request.sendWantList(
+  await self.network.request.sendWantList(
     blockPeer.id, addresses, wantType = WantType.WantBlock
   ) # we want this remote to send us a block
   codex_block_exchange_want_block_lists_sent.inc()
 
-proc monitorBlockHandle(
-    b: BlockExcEngine, handle: Future[Block], address: BlockAddress, peerId: PeerId
-) {.async.} =
-  try:
-    discard await handle
-  except CancelledError as exc:
-    trace "Block handle cancelled", address, peerId
-  except CatchableError as exc:
-    warn "Error block handle, disconnecting peer", address, exc = exc.msg, peerId
+proc randomPeer(peers: seq[BlockExcPeerCtx]): BlockExcPeerCtx =
+  Rng.instance.sample(peers)
 
-    # TODO: really, this is just a quick and dirty way of
-    # preventing hitting the same "bad" peer every time, however,
-    # we might as well discover this on or next iteration, so
-    # it doesn't mean that we're never talking to this peer again.
-    # TODO: we need a lot more work around peer selection and
-    # prioritization
+proc downloadInternal(
```

Comment:
Just a though. I am wondering if `downloadInternal` does not better belong to `pendingblocks`. It would perhaps drag with it other things that should be there. For example `RetriesExhaustedError` is defined in `pendingblocks`, yet never used there but here.

Reply:
I originally had this as a template in pendingBlocks, but I opted for moving it here, because the downloading logic should be handled in engine. `RetriesExhausted` is being used in pendingBlocks as part of the type declaration, so I don't see an issue.